velox/exec/tests/TableWriteTest.cpp

```diff
@@ -610,3 +562,122 @@ TEST_F(TableWriteTest, writeNoFile) {
   execute(plan, std::make_shared<core::QueryCtx>(executor_.get()));
   ASSERT_TRUE(fs::is_empty(outputDirectory->path));
 }
+
+TEST_F(TableWriteTest, insertUnpartitionedTableWithExistingTableType) {
+  // When table type is EXISTING and "immutable_partitions" config is set to
+  // true, inserts into such unpartitioned tables are not allowed.
+  //
+  // We assert that an error is thrown in this case.
+
+  auto input = makeVectors(rowType_, 10, 10);
+  auto outputDirectory = TempDirectoryPath::create();
+  auto plan = createInsertPlan(
+      PlanBuilder().values(input),
+      rowType_,
+      outputDirectory->path,
+      {},
+      connector::hive::LocationHandle::TableType::kExisting);
+
+  VELOX_ASSERT_THROW(
+      AssertQueryBuilder(plan)
+          .connectorConfig(
+              kHiveConnectorId, HiveConfig::kImmutablePartitions, "true")
+          .copyResults(pool()),
+      "Unpartitioned Hive tables are immutable.");
+}
+
+TEST_F(TableWriteTest, insertUnpartitionedTableWithNewTableType) {
+  // When table type is NEW, we always return UpdateMode::kNew. In this case
+  // no exception is expected because we are trying to insert rows into a new
+  // table.
+
+  for (auto immutablePartitionsEnabled : {"true", "false"}) {
+    auto input = makeVectors(rowType_, 10, 10);
+    auto outputDirectory = TempDirectoryPath::create();
+    auto plan = createInsertPlan(
+        PlanBuilder().values(input),
+        rowType_,
+        outputDirectory->path,
+        {},
+        connector::hive::LocationHandle::TableType::kNew);
+
+    auto result = AssertQueryBuilder(plan)
+                      .connectorConfig(
+                          kHiveConnectorId,
+                          HiveConfig::kImmutablePartitions,
+                          immutablePartitionsEnabled)
+                      .copyResults(pool());
+
+    assertEqualResults(
+        {makeRowVector({makeConstant<int64_t>(100, 1)})}, {result});
+  }
+}
+
+TEST_F(TableWriteTest, insertUnpartitionedTable) {
```

Comment:
perhaps, appendToUnpartitionedTable

Should this test be combined with an earlier one? We could verify that the query fails if immutable_partitions is true and succeeds if it is false.